overlord/ifacestate/ifacestate.go

```diff
@@ -35,17 +36,90 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
-var noConflictOnConnectTasks = func(task *state.Task) bool {
-	return task.Kind() != "connect" && task.Kind() != "disconnect"
+func noConflictOnConnectTasks(chg *state.Change) func(task *state.Task) bool {
+	checkConflict := func(task *state.Task) bool {
+		if chg != nil && task.Change() != nil {
+			// if same change, then return false (no conflict)
+			return chg.ID() != task.Change().ID()
+		}
+		return task.Kind() != "connect" && task.Kind() != "disconnect"
+	}
+	return checkConflict
+}
+
+var connectRetryTimeout = time.Second * 5
+
+func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slotSnap string) error {
+	for _, chg := range st.Changes() {
+		if chg.Status().Ready() || (change != nil && change.ID() == chg.ID()) {
+			continue
+		}
+		if chg.Kind() == "transition-ubuntu-core" {
+			return fmt.Errorf("ubuntu-core to core transition in progress, no other changes allowed until this is done")
+		}
+	}
+
+	for _, task := range st.Tasks() {
+		k := task.Kind()
+		chg := task.Change()
+
+		// there is no conflict if same change
+		if change != nil && chg != nil && change.ID() == chg.ID() {
+			continue
+		}
+
+		if k == "connect" || k == "disconnect" {
+			continue
+		}
+
+		// task already executed, no conflict
+		if task.Status() == state.DoneStatus || task.Status() == state.ErrorStatus {
+			continue
+		}
+
+		snapsup, err := snapstate.TaskSnapSetup(task)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot obtain snap setup from task: %s", task.Summary())
+		}
+
+		snapName := snapsup.Name()
+		// different snaps - no conflict
+		if snapName != plugSnap && snapName != slotSnap {
+			continue
+		}
+
+		switch k {
+		case "unlink-snap":
+			// snap getting removed - conflict
+			var snap string
+			if snapName == plugSnap {
+				snap = plugSnap
+			} else {
+				snap = slotSnap
+			}
+			return snapstate.ChangeConflictError(snap, task.Change().Kind())
+		case "link-snap":
+			// snap getting installed/refreshed - temporary conflict, retry later
+			return &state.Retry{After: connectRetryTimeout}
```

Comment:
this looks wrong if we are called by Connect, no? we want a straight conflict there,  shouldn't we behave a bit differently in the auto-connect vs not case

Reply:
Agreed, done (restored old behavior, basically).